WD-Firmware/demo/demo_comrv_baremetal.c

```diff
@@ -0,0 +1,179 @@
+#include "comrv_api.h"
+
+extern void* __OVERLAY_STORAGE_START__ADDRESS__;
+
+#ifdef D_COMRV_FW_INSTRUMENTATION
+comrvInstrumentationArgs_t g_stInstArgs;
+#endif /* D_COMRV_FW_INSTRUMENTATION */
+
+#define OVL_OverlayFunc0 _OVERLAY_
+#define OVL_OverlayFunc1 _OVERLAY_
+#define OVL_OverlayFunc2 _OVERLAY_
+#define OVL_benchmark  //_OVERLAY_
+#define OVL_jpegdct  //_OVERLAY_
+
+//extern int OVL_benchmark benchmark(void);
+
+volatile u32_t globalCount = 0;
+volatile u32_t gOverlayFunc0 = 0;
+volatile u32_t gOverlayFunc1 = 0;
+volatile u32_t gOverlayFunc2 = 0;
+
+/* overlay function 2 */
+void OVL_OverlayFunc2 OverlayFunc2(void)
+{
+   gOverlayFunc2+=3;
+}
+
+/* non overlay function */
+__attribute__((noinline)) void NonOverlayFunc(void)
+{
+   globalCount+=1;
+   OverlayFunc2();
+   globalCount+=2;
+}
+
+/* overlay function 1 */
+void OVL_OverlayFunc1 OverlayFunc1(void)
+{
+   gOverlayFunc1+=3;
+   NonOverlayFunc();
+   gOverlayFunc1+=4;
+}
+
+/* overlay function 0 */
+void OVL_OverlayFunc0 OverlayFunc0(void)
+{
+   gOverlayFunc0+=1;
+   OverlayFunc1();
+   gOverlayFunc0+=2;
+}
+
+void demoStart(void)
+{
+   comrvInitArgs_t stComrvInitArgs = { 1 };
+
+   comrvInit(&stComrvInitArgs);
+
+   globalCount+=1;
+   OverlayFunc0();
+   //benchmark();
+   globalCount+=2;
+   /* verify function calls where completed successfully */
+   if (globalCount != 6 || gOverlayFunc0 != 3 ||
+       gOverlayFunc1 != 7 || gOverlayFunc2 != 3)
+   {
+      /* loop forever */
+      while(1);
+   }
+}
+
+/**
+* memory copy hook
+*
+* @param  none
+*
+* @return none
+*/
+void comrvMemcpyHook(void* pDest, void* pSrc, u32_t sizeInBytes)
+{
+   u32_t loopCount = sizeInBytes/(sizeof(u32_t)), i;
+   /* copy dwords */
+   for (i = 0; i < loopCount ; i++)
+   {
+      *((u32_t*)pDest + i) = *((u32_t*)pSrc + i);
+   }
+   loopCount = sizeInBytes - (loopCount*(sizeof(u32_t)));
+   /* copy bytes */
+   for (i = (i-1)*(sizeof(u32_t)) ; i < loopCount ; i++)
+   {
+      *((u08_t*)pDest + i) = *((u08_t*)pSrc + i);
+   }
+}
+
+/**
+* load overlay group hook
+*
+* @param pLoadArgs - refer to comrvLoadArgs_t for exact args
+*
+* @return loaded address or NULL if unable to load
+*/
+void* comrvLoadOvlayGroupHook(comrvLoadArgs_t* pLoadArgs)
+{
+   comrvMemcpyHook(pLoadArgs->pDest, (u08_t*)&__OVERLAY_STORAGE_START__ADDRESS__ + pLoadArgs->uiGroupOffset, pLoadArgs->uiSizeInBytes);
+   return pLoadArgs->pDest;
+}
+
+/**
+* notification hook
+*
+* @param  pInstArgs - pointer to instrumentation arguments
+*
+* @return none
+*/
+#ifdef D_COMRV_FW_INSTRUMENTATION
+void comrvInstrumentationHook(const comrvInstrumentationArgs_t* pInstArgs)
+{
+   g_stInstArgs = *pInstArgs;
+}
+#endif /* D_COMRV_FW_INSTRUMENTATION */
+
+/**
+* error hook
+*
+* @param  pErrorArgs - pointer to error arguments
+*
+* @return none
+*/
+void comrvErrorHook(const comrvErrorArgs_t* pErrorArgs)
+{
+   comrvStatus_t stComrvStatus;
+   comrvGetStatus(&stComrvStatus);
+   /* we can't continue so loop forever */
+   while (1);
```

Comment:
to illuminate compiler "purging" or optimization, please add "nops" inside the loop (can be new Macro for endless loop)

Reply:
done in 6c72e9f